README.md
Wrangler
========

A file manager with sophisticated metadata handling capabilities

## DESCRIPTION

Extended file attributes are a very versatile and powerful extension of traditional
file system semantics. Yet, most end-user applications ignore xattribs, or in cases
where an app choses to make xattribs accessible for the average user, the actual
user-interface is hidden in "file properties" sub menus, or cumbersome to use.

Wrangler is a "file manager"-like application that puts file metadata first, offering
xattribs and other metadata alongside traditional metadata (size,type,mtime,...)
in all of it's views. The application was designed to browse and manage large collections
of multimedia content files, digital assets, and their associated metadata.

A modular application-layout in combination with a Plugin facility makes Wrangler
adaptable to most workflows or work environments. The central file-browser can be
complemented with a Navbar and Sidebar widget, for comfortable browsing, or with
more specialised multimedia widgets: Wrangler's image/video Previewer or the Metadata-Editor.
If you try only one feature, and you haven't used xattribs until now, then test
the Previewer. It reads preview-thumbnails embedded into JPEGs and takes the lag
out of browsing large image file collections - without maintaining an additional
database.

Wrangler is not meant as a replacement for your primary file manager.
Wrangler is primarily a metadata handling
application, while it also offers the interface and most functionalities commonly
found in file-managers for navigating filesystems and selecting files. But if you
end up using Wrangler for everyday file browsing, that's okay with us.

_Please note:_

This here is only a short github placeholder README. More information about
how to install and use Wrangler can be found in the POD embedded in the source code.
So, please hop over to:

- [Wrangler's Documentation](http://search.cpan.org/perldoc?Wrangler).
- [Wrangler's Official Homepage](http://www.clipland.com/wrangler)

## SCREENSHOTS

<div>
<a href="https://raw.github.com/clipland/wrangler/master/screenshot1.png"><span><img src="https://raw.github.com/clipland/wrangler/master/screenshot1_small.png" width="400" height="300" alt="Screenshot 1" style="border: 1px solid #888;" /></span></a>
<a href="https://raw.github.com/clipland/wrangler/master/screenshot2.png"><span><img src="https://raw.github.com/clipland/wrangler/master/screenshot2_small.png" width="400" height="300" alt="Screenshot 1" style="border: 1px solid #888;" /></span></a>
</div>

## INSTALLATION

via CPAN (official releases):

    sudo cpan -i Wrangler

from command-line (latest changes, if any):

    wget https://github.com/clipland/wrangler/archive/master.tar.gz
    tar xvf master.tar.gz
    cd wrangler-master
    perl Makefile.PL
    make
    make test
    sudo make install

via .deb file:

    go to http://www.clipland.com/wrangler
    download .deb file
    use gdebi or similar to install
    install File::ExtAttr via CPAN (as long as there's no libfile-extattr-perl package)


## AUTHOR

Clipland GmbH, [clipland.com](http://www.clipland.com/)

## COPYRIGHT

Copyright 2009-2014 Clipland GmbH. All rights reserved.

## LICENSE

Wrangler is dual-licensed under the _Wrangler Non-Commercial License_ for private,
non-commercial use, free-of-charge; and under a purchasable license for commercial,
institutional and educational use. Please contact Clipland at http://www.clipland.com/wrangler
to buy commercial licenses.

Please note that Wrangler's license keeps it from being [officially "open source" software](http://opensource.org/faq#avoid-unapproved-licenses).
Nor is it GNU "free software", as it permits only one (freedom 2) of the [four freedoms](http://www.gnu.org/philosophy/free-sw.html).

Wrangler falls into Debian's non-free software category, as the Wrangler Licenses
do not allow derived works, which would be rule 3 of the [Debian Free Software Guidelines (DFSG)](https://www.debian.org/doc/debian-policy/ch-archive.html).

Wrangler relies on a number of Perl modules and the WxWidgets toolkit. If you are
interested in the licensing and copyright status of these modules, please have a
look at Makefile.PL which contains some notes.


Changes
Revision history for Wrangler

2.14	2014-06-23
	- Previewer: fix 'no thumb' rendering for newer Wx
	- Sidebar: look for gtk bookmarks file in gtk3 location
	- FileSystem::Linux: decode strings to perl-internal
	- File::Slurp replaced with Path:Tiny, as per RT#95678
	- KeywordingTool: nonsensical dependencies removed
	- Previewer: lazy helper probing for libav-tools
	- debian control now lists libav-tools as dependency
	- tests: adapted to gvfs-trash behaviour change

2.13	2014-05-12
	- currently unused dependencies XML::Simple and
	  Crypt::SSLeay dropped from Makefile.PL
	- RePopulate triggered by dir change monitoring now less
	  annoying after renames and deletes
	- new Dialog KeywordingTool
	- FileBrowser: ListCtrl instead of ListView
	- changed deb package build system

2.12	2014-04-02
	- FileBrowser: Paste operations now handle utf8
	  paths, and give UI feedback on error
	- FileBrowser reclaims keyboard-focus on Populate
	- FileBrowser: very basic UI feedback on copy/move
	  file operations
	- Linux: move() now uses system 'mv' command
	- Previewer: constructs no_preview bitmap internally
	- changed CLI parsing, Wrangler now opens a
	  supplied path instead of always opening with cwd
	- Default config now has working shortcuts set so
	  the example "value-shortcuts" work
	- Enter license key via GUI
	- ListingToText is back

2.11	2014-03-19 (first public release)
	- CustomizableEditor is back, on the 2.x branch
	  as FormEditor, with many improvements, switchable
	  layouts, UI for adding/removing fields, etc.
	- MultiRename dialog ported to 2.x branch
	- Previewer displays video thumbnails again
	- FileBrowser: configurable columns complete,
	  sorting remembered
	- ColourLabels Plugin
	- Layers provides new field "Filesystem::Xattr"
	- Value Shortcuts facility ported to 2.x branch
	- Settings UI complete and refined
	- accepts (--debug 1) for debug output to STDOUT
	- silenced/configurable debug output
	- reworked Plugin API (still WIP)
	- stability fixes
	- app-icon-set completed, optional scalable
	- polished POD, with screenshots

2.10	2014-01-18 (unreleased)
	- MetadataThumb is back, on the 2.x branch as
	  Previewer, with fixes
	- preparation for dynamic layouting
	- Navbar with history and clickable path
	- Sidebar with icons
	- configurable, MIME-Type based viewers facility
	- new Settings UI, more config bindings 
	- Debian .deb file pipeline working
	- new Images.pm build process
	- License and Changes embedded as modules

2.07	2013-12-15 (unreleased)
	- initial PubSub
	- basic Sidebar
	- OO-interface dropped on selected subs for speed
	- ListCtrl FileBrowser mainly done, restored
	  configurable columns, zebra-striping, highlighting
	- new logo, new internal image/icon handling
	- ported tests

2.02	2013-10-30 (unreleased)
	- Wrangler::Config rewrite
	- implementation of FileBrowser as a TreeList halted
	- configurable widget layout
	- new PluginManager

2.0	2013-07-01 (unreleased)
	- start of new major version, branch/rewrite
 	- switch from YAML to JSON for settings
	- new XDG config file location
	- removed sidecar-file handling for now
	- many widgets and dialogs not yet ported to 2.x
	- new underlying file i/o
	- removed separate handling via Metadata.pm
	- FileSystem::Layers
	- native file/folder/icons
	- new, WIP Plugin API
	- events replace hardcoded bindings
	- new module layout
	- better separation of WxWidgets and Wrangler's
	  internal data structures
	- CPAN bundling and proper Licensing

1.36	2013-04-28 (unreleased)
	- fixed video preview and thumbnailing
	- optimized onpaint/redraws
	- stability: various crashers fixed

1.24	2012-02-07 (unreleased)
	- Layered VFS and 'mounting' services
	- Dialog::AdjustTimestamp
	- switchable Metadata editors: CustomizableEditor
	  and MetadataPanel/MetadataTree
	- internal metadata API via Metadata.pm

1.17	2011-08-16 (unreleased)
	- Dialog::PlaylistEditor (ASX, XML, M3U, ..)
	- FileBrowser zooming/fontsize
	- export listings as text
	- initial metadata mapping tool OneToMany
	- more filetypes, WebM etc.
	- alternative metadata editor CustomizableEditor
	- Previewer with slider: pos and time
	- internal/native "viewers" for Images, Video
	  and Audio
	- insert editable text-templates via configurable
	  keyboard shortcuts
	- PAR packaging
	- cross platform fixes for Windows and Linux

1.0	2010-11-01 (unreleased)
	- major rewrites in many areas
	- FileBrowser is not a WxGrid anymore
	- FileBrowser as HTML window prototype
	- MetadataPanel display and editor
	- MetadataThumb previewer
	- new column sorting
	- fixed utf8 (filename) handling
	- proper Settings UI
	- sidecar-file handling and in-listing indenting
	- Accounts manager
	- Dialog::MultiRename
	- i18n

0.5	2009-02-15 (unreleased)
	- dropped tabbed main interface
	- metadata handling logic changed
	- switch to sqlite for backend db
	- l18n internationalization ready
	- unified internal module and data structure
	- file-browsing rewritten
	- working trancoding wizard, Convert.pm
	- renaming from 'Video Wrangler' to 'Wrangler'
	  reflecting an universal approach to media asset
	  handling
	- countless bugfixes, changes
	- Db backend dropped, pure xattr from now on
	- Metadata Details display

0.4	2009-02-19 (unreleased)
	- Dialog::Watermark
	- Plugin::Veoh, Plugin::MySpace

0.3	2008-08-13 (unreleased)
	- Plugin::Dailymotion
	- Dialog::Export

0.2	2008-07-04 (unreleased)
	- switch to WxWidgets
	- early Plugin interface
	- initial upload facility

0.1	2008-03-11 (unreleased)
	- initial Tk version
	- interface design prototype
	- Storable-based backend db


lib/Wrangler/Changes.pm
package Wrangler::Changes;
#
# Warning: This module was auto-generated by build_textmodules.pl, manual changes will be lost!
#

use strict;
use warnings;

our $text = "Revision history for Wrangler

2.14	2014-06-23
	- Previewer: fix 'no thumb' rendering for newer Wx
	- Sidebar: look for gtk bookmarks file in gtk3 location
	- FileSystem::Linux: decode strings to perl-internal
	- File::Slurp replaced with Path:Tiny, as per RT#95678
	- KeywordingTool: nonsensical dependencies removed
	- Previewer: lazy helper probing for libav-tools
	- debian control now lists libav-tools as dependency
	- tests: adapted to gvfs-trash behaviour change

2.13	2014-05-12
	- currently unused dependencies XML::Simple and
	  Crypt::SSLeay dropped from Makefile.PL
	- RePopulate triggered by dir change monitoring now less
	  annoying after renames and deletes
	- new Dialog KeywordingTool
	- FileBrowser: ListCtrl instead of ListView
	- changed deb package build system

2.12	2014-04-02
	- FileBrowser: Paste operations now handle utf8
	  paths, and give UI feedback on error
	- FileBrowser reclaims keyboard-focus on Populate
	- FileBrowser: very basic UI feedback on copy/move
	  file operations
	- Linux: move() now uses system 'mv' command
	- Previewer: constructs no_preview bitmap internally
	- changed CLI parsing, Wrangler now opens a
	  supplied path instead of always opening with cwd
	- Default config now has working shortcuts set so
	  the example \"value-shortcuts\" work
	- Enter license key via GUI
	- ListingToText is back

2.11	2014-03-19 (first public release)
	- CustomizableEditor is back, on the 2.x branch
	  as FormEditor, with many improvements, switchable
	  layouts, UI for adding/removing fields, etc.
	- MultiRename dialog ported to 2.x branch
	- Previewer displays video thumbnails again
	- FileBrowser: configurable columns complete,
	  sorting remembered
	- ColourLabels Plugin
	- Layers provides new field \"Filesystem::Xattr\"
	- Value Shortcuts facility ported to 2.x branch
	- Settings UI complete and refined
	- accepts (--debug 1) for debug output to STDOUT
	- silenced/configurable debug output
	- reworked Plugin API (still WIP)
	- stability fixes
	- app-icon-set completed, optional scalable
	- polished POD, with screenshots

2.10	2014-01-18 (unreleased)
	- MetadataThumb is back, on the 2.x branch as
	  Previewer, with fixes
	- preparation for dynamic layouting
	- Navbar with history and clickable path
	- Sidebar with icons
	- configurable, MIME-Type based viewers facility
	- new Settings UI, more config bindings 
	- Debian .deb file pipeline working
	- new Images.pm build process
	- License and Changes embedded as modules

2.07	2013-12-15 (unreleased)
	- initial PubSub
	- basic Sidebar
	- OO-interface dropped on selected subs for speed
	- ListCtrl FileBrowser mainly done, restored
	  configurable columns, zebra-striping, highlighting
	- new logo, new internal image/icon handling
	- ported tests

2.02	2013-10-30 (unreleased)
	- Wrangler::Config rewrite
	- implementation of FileBrowser as a TreeList halted
	- configurable widget layout
	- new PluginManager

2.0	2013-07-01 (unreleased)
	- start of new major version, branch/rewrite
 	- switch from YAML to JSON for settings
	- new XDG config file location
	- removed sidecar-file handling for now
	- many widgets and dialogs not yet ported to 2.x
	- new underlying file i/o
	- removed separate handling via Metadata.pm
	- FileSystem::Layers
	- native file/folder/icons
	- new, WIP Plugin API
	- events replace hardcoded bindings
	- new module layout
	- better separation of WxWidgets and Wrangler's
	  internal data structures
	- CPAN bundling and proper Licensing

1.36	2013-04-28 (unreleased)
	- fixed video preview and thumbnailing
	- optimized onpaint/redraws
	- stability: various crashers fixed

1.24	2012-02-07 (unreleased)
	- Layered VFS and 'mounting' services
	- Dialog::AdjustTimestamp
	- switchable Metadata editors: CustomizableEditor
	  and MetadataPanel/MetadataTree
	- internal metadata API via Metadata.pm

1.17	2011-08-16 (unreleased)
	- Dialog::PlaylistEditor (ASX, XML, M3U, ..)
	- FileBrowser zooming/fontsize
	- export listings as text
	- initial metadata mapping tool OneToMany
	- more filetypes, WebM etc.
	- alternative metadata editor CustomizableEditor
	- Previewer with slider: pos and time
	- internal/native \"viewers\" for Images, Video
	  and Audio
	- insert editable text-templates via configurable
	  keyboard shortcuts
	- PAR packaging
	- cross platform fixes for Windows and Linux

1.0	2010-11-01 (unreleased)
	- major rewrites in many areas
	- FileBrowser is not a WxGrid anymore
	- FileBrowser as HTML window prototype
	- MetadataPanel display and editor
	- MetadataThumb previewer
	- new column sorting
	- fixed utf8 (filename) handling
	- proper Settings UI
	- sidecar-file handling and in-listing indenting
	- Accounts manager
	- Dialog::MultiRename
	- i18n

0.5	2009-02-15 (unreleased)
	- dropped tabbed main interface
	- metadata handling logic changed
	- switch to sqlite for backend db
	- l18n internationalization ready
	- unified internal module and data structure
	- file-browsing rewritten
	- working trancoding wizard, Convert.pm
	- renaming from 'Video Wrangler' to 'Wrangler'
	  reflecting an universal approach to media asset
	  handling
	- countless bugfixes, changes
	- Db backend dropped, pure xattr from now on
	- Metadata Details display

0.4	2009-02-19 (unreleased)
	- Dialog::Watermark
	- Plugin::Veoh, Plugin::MySpace

0.3	2008-08-13 (unreleased)
	- Plugin::Dailymotion
	- Dialog::Export

0.2	2008-07-04 (unreleased)
	- switch to WxWidgets
	- early Plugin interface
	- initial upload facility

0.1	2008-03-11 (unreleased)
	- initial Tk version
	- interface design prototype
	- Storable-based backend db
";

1;